### PR TITLE
feat: Skip layers due to size mismatch when loading checkpoint

### PR DESCRIPTION
**Pull request by theissenhelen**

### CODE REVIEW

**Comment by gabrieloks** [comment]
Tried this on a n320 training started from an o96 checkpoint and works very well!

`
[2024-11-12 11:01:29,199][anemoi.training.train.train][INFO] - Skipping layer model.model.latlons_data due to size mismatch: torch.Size([40320, 4]) vs torch.Size([542080, 4]).
[2024-11-12 11:01:29,199][anemoi.training.train.train][INFO] - Skipping layer loss.weights due to size mismatch: torch.Size([40320]) vs torch.Size([542080]).
[2024-11-12 11:01:29,199][anemoi.training.train.train][INFO] - Skipping layer metrics.weights due to size mismatch: torch.Size([40320]) vs torch.Size([542080]).
`

**Review by gabrieloks — DISMISSED** [review]
LGTM

**Comment by HCookie** [comment]
I feel as if this change may require a changelog update

**Comment by JPXKQX** [comment]
Would it make sense to overwrite the `load_from_checkpoint` method with this piece of code?

File: src/anemoi/training/train/train.py [hunk 1]
```diff
@@ -144,10 +144,29 @@ def model(self) -> GraphForecaster:
...
             "statistics": self.datamodule.statistics,
         }
+        model = GraphForecaster(**kwargs)
+
         if self.load_weights_only:
             LOGGER.info("Restoring only model weights from %s", self.last_checkpoint)
-            return GraphForecaster.load_from_checkpoint(self.last_checkpoint, **kwargs)
-        return GraphForecaster(**kwargs)
+            checkpoint = torch.load(self.last_checkpoint)
+            checkpoint_state_dict = checkpoint["state_dict"]
```
Comment (HCookie) [inline src/anemoi/training/train/train.py:152]: It may be good to have this behaviour to be toggleable, so a config is needed to be set to enable this size checking. Otherwise a user may attempt to load a checkpoint of an invalid model, and every layer just skipped with only logs as awareness.

Comment (ssmmnn11) [reply src/anemoi/training/train/train.py:152]: Yes, that makes sense. 

Btw, do we have a check that a checkpoint we specify when warm starting actually exist? at least at some point the model just complaint that it can't find the checkpoint and then does a cold start.

**Comment by havardhhaugen** [comment]
You can loop over only the layers you want to replace using model.named_parameters (this does not contain the lat-lon data)
This is the implementation we used to do transfer-learning with the aifs-mono setup:
```
T = AIFSTrainer(config)

if T.load_weights_only:
    #print("start from checkpoint", T.start_from_checkpoint)
    print('Loading weights only from checkpoint: {}'.format(T.last_checkpoint))
    model_to_train_from = torch.load(T.last_checkpoint, 'cpu')
    for name, param in T.model.named_parameters():
        param.data = model_to_train_from['state_dict'][name].data
```
This should work inside Trainer too as long as you have access to GraphForecaster.named_parameters

File: src/anemoi/training/train/train.py (hunk 1 above)
Comment (HCookie) [reply src/anemoi/training/train/train.py:152]: I'm not sure, I would imagine that it should fail when it can't find the checkpoint

**Comment by ssmmnn11** [comment]
I don't have time currently to do the changes and testing. If someone wants to take over, please do.

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-training?pullRequest=140) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-training?pullRequest=140) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: ssmmnn11<br/>:white_check_mark: gabrieloks<br/>:x: MOLDOVAN Gabriel Ionut<br/><hr/>**MOLDOVAN Gabriel Ionut** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-training?pullRequest=140) it.</sub>

**Review by HCookie — APPROVED** [review]

**Comment by gabrieloks** [comment]
Closing since duplicate
